ebpf/xdp_lb.c

```diff
@@ -0,0 +1,173 @@
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+#define KBUILD_MODNAME "foo"
+#include <stddef.h>
+#include <linux/bpf.h>
+
+#include <linux/in.h>
+#include <linux/if_ether.h>
+#include <linux/if_packet.h>
+#include <linux/if_vlan.h>
+#include <linux/ip.h>
+#include <linux/ipv6.h>
+#include <linux/tcp.h>
+#include <linux/udp.h>
+#include "bpf_helpers.h"
+
+#include "hash_func01.h"
+
+#define LINUX_VERSION_CODE 263682
+
+/* Hashing initval */
+#define INITVAL 15485863
+
+/* Increase CPUMAP_MAX_CPUS if ever you have more than 64 CPUs */
+#define CPUMAP_MAX_CPUS     64
+
+struct vlan_hdr {
+    __u16	h_vlan_TCI;
+    __u16	h_vlan_encapsulated_proto;
+};
+
+/* Special map type that can XDP_REDIRECT frames to another CPU */
+struct bpf_map_def SEC("maps") cpu_map = {
+    .type		= BPF_MAP_TYPE_CPUMAP,
+    .key_size	= sizeof(__u32),
+    .value_size	= sizeof(__u32),
+    .max_entries	= CPUMAP_MAX_CPUS,
+};
+
+struct bpf_map_def SEC("maps") cpus_available = {
+    .type		= BPF_MAP_TYPE_ARRAY,
+    .key_size	= sizeof(__u32),
+    .value_size	= sizeof(__u32),
+    .max_entries	= CPUMAP_MAX_CPUS,
+};
+
+struct bpf_map_def SEC("maps") cpus_count = {
+    .type		= BPF_MAP_TYPE_ARRAY,
+    .key_size	= sizeof(__u32),
+    .value_size	= sizeof(__u32),
+    .max_entries	= 1,
+};
+
+static int __always_inline filter_ipv4(struct xdp_md *ctx, void *data, __u64 nh_off, void *data_end)
+{
+    struct iphdr *iph = data + nh_off;
+    __u32 key0 = 0;
+    __u32 cpu_dest;
+    __u32 *cpu_max = bpf_map_lookup_elem(&cpus_count, &key0);
+    __u32 *cpu_selected;
+    __u32 cpu_hash;
+
+    if ((void *)(iph + 1) > data_end)
+        return XDP_PASS;
+
+    /* IP-pairs hit same CPU */
+    cpu_hash = iph->saddr + iph->daddr;
+    cpu_hash = SuperFastHash((char *)&cpu_hash, 4, INITVAL);
+
+    if (cpu_max && *cpu_max) {
+        cpu_dest = cpu_hash % *cpu_max;
+        cpu_selected = bpf_map_lookup_elem(&cpus_available, &cpu_dest);
+        if (!cpu_selected)
+            return XDP_ABORTED;
+        cpu_dest = *cpu_selected;
+        return bpf_redirect_map(&cpu_map, cpu_dest, 0);
+    } else {
+        return XDP_PASS;
+    }
+}
+
+static int __always_inline filter_ipv6(struct xdp_md *ctx, void *data, __u64 nh_off, void *data_end)
+{
+    struct ipv6hdr *ip6h = data + nh_off;
+    __u32 key0 = 0;
+    __u32 cpu_dest;
+    int *cpu_max = bpf_map_lookup_elem(&cpus_count, &key0);
```

Comment:
should __u32 be used here like in the filter_ipv4 code?

Reply:
Yes, indeed.